File: src/days/day8/1.go
package main

import (
	"fmt"
	"bufio"
	"os"
	"strings"
	"strconv"
	"errors"
)

type Instruction struct {
    Opt string
    Steps int
}

func readInstruction(path string) []Instruction {
	file, _ := os.Open("./in")
	defer file.Close()
	scanner := bufio.NewScanner(file)

	var instructions []Instruction
	for scanner.Scan() {
        instruction := strings.Fields(scanner.Text())
        if len(instruction) == 2 {
            opt := instruction[0]
            val, _ := strconv.Atoi(instruction[1])
            instructions = append(instructions, Instruction{opt, val})
        }
	}
	return instructions
}

func run(instruction Instruction, curIdx int, accumulator int) (int, int) {
    switch opt := instruction.Opt; opt {
    	case "acc":
    	    return curIdx+1, accumulator+instruction.Steps
    	case "jmp":
    	    return curIdx+instruction.Steps, accumulator
    	case "nop":
    	    return curIdx+1, accumulator
    	default:
    		fmt.Printf("%s.\n", opt)
    	}
    return 0, 0
}


func runProgram(instructions []Instruction) (int, error) {
    curIdx := 0
    accumulator := 0
    seen := map[int]bool{}
    for  {
        if _, ok := seen[curIdx]; !ok && curIdx < len(instructions) {
            seen[curIdx]=true
            instruction := instructions[curIdx]
            curIdx, accumulator = run(instruction, curIdx, accumulator)
        } else {
            break
        }
    }
    if curIdx < len(instructions) {
        return accumulator, errors.New("Stop")
    }
    return accumulator, nil
}


func swapOpt(opt string) string {
    if opt == "jmp" {
        return "nop"
    } else if opt == "nop" {
        return "jmp"
    }
    fmt.Println("you should not see this")
    return ""
}


func execPart1(instructions []Instruction) int {
	acc, err := runProgram(instructions)
	if err != nil {
	    fmt.Println(acc)
	    return acc
	}
	return 0
}


func execPart2(instructions []Instruction) int {
    for i, instruction := range instructions {
        if instruction.Opt != "acc" {
            cur := instruction
            instruction.Opt = swapOpt(instruction.Opt)
            instructions[i] = instruction
            acc, err := runProgram(instructions)
            if err != nil {
                instructions[i] = cur
            } else {
                fmt.Println(acc)
                return acc
            }
        }
    }
    return 0
}


func main() {
	instructions := readInstruction("./in")
	execPart1(instructions)
	execPart2(instructions)
}

File: src/days/day8/1.py
def read_instructions():
    f = open('./in', 'r')
    content = [item for item in f.read().strip().split("\n") ]
    f.close()
    return content

def normalize_inst(inst):
    a, b = inst.split( )
    return a, int(b)

def perform_opt(inst, cur, acc):
    opt, val = normalize_inst(inst)
    return {
        'nop': lambda c, a : (c+1, a),
        'jmp': lambda c, a : (c+val, a),
        'acc': lambda c, a : (c+1, a+val),
    }[opt](cur, acc)

def run_programming(instructions):
    cur_idx, accumulator, seen= 0, 0, set()
    while cur_idx not in seen and cur_idx < len(instructions):
        seen.add(cur_idx)
        cur_idx, accumulator = perform_opt(instructions[cur_idx], cur_idx, accumulator)
    return accumulator, cur_idx == len(instructions)

def get_swap_opt(opt):
    return {
        'jmp': 'nop',
        'nop': 'jmp',
        'acc': 'acc',
    }[opt]

def swapped_opt(opt, instruction):
    return instruction.replace(opt, get_swap_opt(opt))

def exec_part1(instructions):
    acc, correct = run_programming(instructions)
    return acc

def exec_part2(instructions):
    n=len(instructions)
    for i in range(n):
        opt, _ = normalize_inst(instructions[i])
        current_instruction = instructions[i]
        instructions[i] = swapped_opt(opt, current_instruction)
        acc, correct = run_programming(instructions)
        if correct:
            return acc
        else:
            instructions[i] = current_instruction
    return None


instruction_list=read_instructions()

print("Part 1:", exec_part1(instruction_list))
print("Part 2:", exec_part2(instruction_list))
